src/AudioCaptureImpl_WASAPI.cpp

```diff
@@ -1,25 +1,631 @@
+        {
+            poco_error_f1(_logger, "IAudioClient::GetMixFormat returned non-float sample format: 0x%04?x", pwfx->wFormatTag);
+            CoTaskMemFree(pwfx);
+            return false;
+        }
+        else if (!IsEqualGUID(KSDATAFORMAT_SUBTYPE_IEEE_FLOAT, extensibleFormat->SubFormat))
+        {
+            poco_error_f1(_logger, "IAudioClient::GetMixFormat returned non-float extensible sub format: 0x%04?x", extensibleFormat->SubFormat);
+            CoTaskMemFree(pwfx);
+            return false;
+        }
+    }
+
+    _channels = pwfx->nChannels;
+
+    // Can't use event-driven processing in loopback mode, but as we
+    // get a "fill buffer" request before rendering each frame, this isn't
+    // really necessary anyway.
+    result = _audioClient->Initialize(
+        AUDCLNT_SHAREMODE_SHARED,
+        useLoopback ? AUDCLNT_STREAMFLAGS_LOOPBACK : 0,
+        hnsDefaultDevicePeriod,
+        0,
+        pwfx,
+        nullptr);
+
+    CoTaskMemFree(pwfx);
+
+    if (FAILED(result))
+    {
+        poco_error_f1(_logger, "IAudioClient->Initialize failed: result = 0x%08?x", result);
+        return false;
+    }
+
+    // activate an IAudioCaptureClient
+    result = _audioClient->GetService(
+        __uuidof(IAudioCaptureClient),
+        (void**) &_audioCaptureClient);
+
+    if (FAILED(result))
+    {
+        poco_error_f1(_logger, "IAudioClient->GetService failed: result = 0x%08?x", result);
+        return false;
+    }
+
+    result = _audioClient->Start();
+    if (FAILED(result))
+    {
+        poco_error_f1(_logger, "IAudioClient->Start failed: result = 0x%08?x", result);
+        return false;
+    }
+
+    return true;
+}
+
+void AudioCaptureImpl::CloseAudioDevice(IMMDevice* device)
+{
+    poco_trace(_logger, "Stopping audio client.");
+    _audioClient->Stop();
+
+    if (_audioCaptureClient)
+    {
+        poco_trace(_logger, "Releasing audio capture client.");
+        _audioCaptureClient->Release();
+        _audioCaptureClient = nullptr;
+    }
+    if (_audioClient)
+    {
+        poco_trace(_logger, "Releasing audio client.");
+        _audioClient->Release();
+        poco_trace(_logger, "Audio client released.");
+        _audioClient = nullptr;
+    }
+
+    if (device)
+    {
+        poco_trace(_logger, "Releasing audio device.");
+        device->Release();
+    }
+}
+
+void AudioCaptureImpl::CaptureThread()
+{
+    poco_debug(_logger, "Audio capture thread starting.");
+
+    HRESULT result = CoInitializeEx(nullptr, COINIT_MULTITHREADED);
+
+    if (FAILED(result))
+    {
+        poco_error_f1(_logger, "CoInitializeEx() failed: result = 0x%08?x", result);
+        throw std::bad_alloc();
+    }
+
+    IMMDeviceEnumerator* enumerator{GetDeviceEnumerator()};
+
+    if (enumerator == nullptr)
+    {
+        CoUninitialize();
+        throw std::bad_alloc();
+    }
+
+    poco_trace(_logger, "Registering device callbacks.");
+    enumerator->RegisterEndpointNotificationCallback(this);
```

Comment:
It's quite important to register the event callbacks in the same thread the capturing happens, as it can cause deadlocks and other issues. It's also important to not block in the callbacks, that's why I've used atomic flags to message the capture thread to restart on the next loop/frame.